packages/std/src/coins.rs

```diff
@@ -0,0 +1,346 @@
+        if coin.amount.is_zero() {
+            return Ok(());
+        }
+
+        let amount = self.0.entry(coin.denom).or_insert_with(Uint128::zero);
+        *amount = amount.checked_add(coin.amount)?;
+        Ok(())
+    }
+
+    /// Adds the given coins to the collection.
+    /// This takes anything that yields `(denom, amount)` tuples when iterated over.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use cosmwasm_std::{Coin, Coins, coin};
+    ///
+    /// let mut coins = Coins::default();
+    /// let new_coins: Coins = coin(123u128, "ucosm").try_into()?;
+    /// coins.extend(new_coins.to_vec())?;
+    /// assert_eq!(coins, new_coins);
+    /// # cosmwasm_std::StdResult::Ok(())
+    /// ```
+    pub fn extend<C>(&mut self, others: C) -> StdResult<()>
+    where
+        C: IntoIterator<Item = Coin>,
+    {
+        for c in others {
+            self.add(c)?;
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::coin;
+
+    /// Sort a Vec<Coin> by denom alphabetically
+    fn sort_by_denom(vec: &mut [Coin]) {
+        vec.sort_by(|a, b| a.denom.cmp(&b.denom));
+    }
+
+    /// Returns a mockup Vec<Coin>. In this example, the coins are not in order
+    fn mock_vec() -> Vec<Coin> {
+        vec![
+            coin(12345, "uatom"),
+            coin(69420, "ibc/1234ABCD"),
+            coin(88888, "factory/osmo1234abcd/subdenom"),
+        ]
+    }
+
+    /// Return a mockup Coins that contains the same coins as in `mock_vec`
+    fn mock_coins() -> Coins {
+        let mut coins = Coins::default();
+        for coin in mock_vec() {
+            coins.add(coin).unwrap();
+        }
+        coins
+    }
+
+    #[test]
+    fn casting_vec() {
+        let mut vec = mock_vec();
+        let coins = mock_coins();
+
+        // &[Coin] --> Coins
+        assert_eq!(Coins::try_from(vec.as_slice()).unwrap(), coins);
+        // Vec<Coin> --> Coins
+        assert_eq!(Coins::try_from(vec.clone()).unwrap(), coins);
+
+        sort_by_denom(&mut vec);
+
+        // &Coins --> Vec<Coins>
+        // NOTE: the returned vec should be sorted
+        assert_eq!(coins.to_vec(), vec);
+        // Coins --> Vec<Coins>
+        // NOTE: the returned vec should be sorted
+        assert_eq!(coins.into_vec(), vec);
+    }
+
+    #[test]
+    fn casting_str() {
+        // not in order
+        let s1 = "88888factory/osmo1234abcd/subdenom,12345uatom,69420ibc/1234ABCD";
+        // in order
+        let s2 = "88888factory/osmo1234abcd/subdenom,69420ibc/1234ABCD,12345uatom";
+
+        let coins = mock_coins();
+
+        // &str --> Coins
+        // NOTE: should generate the same Coins, regardless of input order
+        assert_eq!(Coins::from_str(s1).unwrap(), coins);
+        assert_eq!(Coins::from_str(s2).unwrap(), coins);
+
+        // Coins --> String
+        // NOTE: the generated string should be sorted
+        assert_eq!(coins.to_string(), s2);
+    }
```

Comment:
Can you add a test case for `Coins::from_str("")`?

Reply:
Now I'm wondering: Do we want to parse `""` into an empty `Coins` instance or error?

Reply:
An empty Coins is consistent IMO. `Coins::default().to_string()` is an empty string, right (do we have a test for that?)

Reply:
We don't have a test for that yet, but I was about to add one. I agree on the consistency